libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import * as signInConstants from '../signInConstants';
+import { ExtendedUserTokenProvider } from '../extendedUserTokenProvider';
+import { Middleware } from '../middlewareSet';
+import { Storage, StoreItem } from '../storage';
+import { TurnContext } from '../turnContext';
+
+import {
+    ActivityTypes,
+    StatusCodes,
+    TokenExchangeInvokeRequest,
+    TokenExchangeInvokeResponse,
+    TokenResponse,
+} from 'botframework-schema';
+
+function getStorageKey(context: TurnContext): string {
+    const activity = context.activity;
+
+    const channelId = activity.channelId;
+    if (!channelId) {
+        throw new Error('invalid activity-missing channelId');
+    }
+
+    const conversationId = activity.conversation?.id;
+    if (!conversationId) {
+        throw new Error('invalid activity-missing Conversation.Id');
```

Comment:
nit: Inconsistent error message formatting. Can you align on the format used and clean up the casing on references to members?